### PR TITLE
Add scripts and examples of TLS features

### PR DESCRIPTION
**Pull request by dragonly**
### What problem does this PR solve?

Add a hack script to support generating certs and configs using `cfssl` for docs [Enable TLS between TiDB Components](https://docs.pingcap.com/tidb-in-kubernetes/stable/enable-tls-between-components) and [Enable TLS for the MySQL Client](https://docs.pingcap.com/tidb-in-kubernetes/stable/enable-tls-for-mysql-client)

### What is changed and how does it work?

`hack/gen-tls-certs.sh` can be used to generate all the needed configs and certs utilizing `cfssl`. It will generate all the stuff into the `output` dir.
`examples/basic-tls` contains example `TidbCluster` and `TidbMonitor` CR which enables TLS between `TidbCluster` components.

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [x] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [x] Need to cherry-pick to the release branch 
- [x] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
NONE
```

### CODE REVIEW

**Review by lonng — COMMENTED** [review]
Please add some documentation for guiding users on how to use these scripts.

**Comment by handlerww** [comment]
We can enrich the contributing guiding documents for these tools and scripts.

**Comment by dragonly** [comment]
Added some docs, PTAL @lonng @handlerww

File: hack/gen-tls-certs.sh [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+hack::ensure_cfssl
+
+echo "TLS_DIR: $TLS_DIR"
+echo "CA_DIR: $CA_DIR"
+
+if [[ -d $TLS_DIR ]]; then
+    echo "tls dir exists, removing"
+    rm -rf $TLS_DIR
+fi
+
```
Comment (DanielZhangQD) [inline hack/gen-tls-certs.sh:141]: Add a `usage` function to explain the usage as we do in `e2e.sh`?

Comment (dragonly) [reply hack/gen-tls-certs.sh:189]: done

File: hack/gen-tls-certs.sh [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+function gen_component_certs() {
+    local KIND=$1
+    local COMP=$2
+    local COMP_DIR=$TLS_DIR/$KIND/$2
+    local PROFILE=$3
+    echo "generating $COMP server certs"
+    mkdir -p $COMP_DIR
+    pushd $COMP_DIR
+    $CFSSL_BIN print-defaults csr |
+        jq ".hosts=[\"127.0.0.1\",\"::1\",\"${TIDB_CLUSTER}-${COMP}\",\"${TIDB_CLUSTER}-${COMP}.${NAMESPACE}\",\"${TIDB_CLUSTER}-${COMP}.${NAMESPACE}.svc\",\"${TIDB_CLUSTER}-${COMP}-peer\",\"${TIDB_CLUSTER}-${COMP}-peer.${NAMESPACE}\",\"${TIDB_CLUSTER}-${COMP}-peer.${NAMESPACE}.svc\",\"*.${TIDB_CLUSTER}-${COMP}-peer\",\"*.${TIDB_CLUSTER}-${COMP}-peer.${NAMESPACE}\",\"*.${TIDB_CLUSTER}-${COMP}-peer.${NAMESPACE}.svc\"]" |
```
Comment (DanielZhangQD) [inline hack/gen-tls-certs.sh:163]:
```
/home/dan/dev/go/src/github.com/pingcap/tidb-operator/output/tls/cluster/pd /home/dan/dev/go/src/github.com/pingcap/tidb-operator
hack/gen-tls-certs.sh: line 163: jq: command not found
hack/gen-tls-certs.sh: line 164: jq: command not found
```

**Comment by DanielZhangQD** [comment]
@csuzhangxc @handlerww PTAL.

File: hack/gen-tls-certs.sh (hunk 2 above)
Comment (dragonly) [reply hack/gen-tls-certs.sh]: replaced with `$JQ_BIN`

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: hack/gen-tls-certs.sh [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    TIDB_CLUSTER           should match the tidb cluster you want to enable TLS features, defaults to "basic-tls"
+    NAMESPACE              namespace of the tidb cluster, defaults to "default"
+    # following are CA cert configs
+    CA_EXPIRATION          expiration time of CA certs, defaults to "87600h" (10*365 Days)
+    COMMON_NAME            COMMON_NAME(CN) of the CA cert, defaults to "TiDB"
+    ORGANIZATION_UNIT      ORGANIZATION_UNIT(OU) of the CA cert, defaults to "TiDB"
+    ORGANIZATION           ORGANIZATION(O) of the CA cert, defaults to "PingCAP"
+    COUNTRY                COUNTRY(C) of the CA cert, defaults to "CN"
+    LOCALITY               LOCALITY(L) of the CA cert, defaults to "Beijing"
+    STATE_OR_PROVINCE_NAME STATE_OR_PROVINCE_NAME(ST) of the CA cert, defaults to Beijing
```
Comment (csuzhangxc) [inline hack/gen-tls-certs.sh:60]:
```suggestion
    STATE_OR_PROVINCE_NAME STATE_OR_PROVINCE_NAME(ST) of the CA cert, defaults to "Beijing"
```

File: hack/gen-tls-certs.sh [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    echo "tls dir exists, removing"
+    rm -rf $TLS_DIR
+fi
+
+mkdir $TLS_DIR
+
+gen_ca
+
+# generate certs for cluster components
+components=("pd" "tidb" "tikv" "pump" "ticdc" "tiflash")
```
Comment (csuzhangxc) [inline hack/gen-tls-certs.sh:195]: do we need to generated certs for `importer` and `lightning` now?

**Comment by csuzhangxc** [comment]
Do we need a command to delete all secrets after tested?

File: hack/gen-tls-certs.sh (hunk 4 above)
Comment (dragonly) [reply hack/gen-tls-certs.sh:195]: there seems no entries for these 2 components in the doc now?
maybe add that support when doc is ready.

**Comment by dragonly** [comment]
> Do we need a command to delete all secrets after tested?

Thanks for reminding. I can add a `clean` subcommand in the script.

File: examples/basic-tls/tidb-cluster.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    baseImage: pingcap/tidb
+    replicas: 1
+    service:
+      type: NodePort
+    tlsClient:
+      enabled: true
+    config:
+      security:
+        cert-verify-cn:
+          - TiDB
```
[the remark this answers is not in the document]
Comment (dragonly) [reply examples/basic-tls/tidb-cluster.yaml:57]: WDYT @DanielZhangQD

File: hack/gen-tls-certs.sh (hunk 4 above)
Comment (csuzhangxc) [reply hack/gen-tls-certs.sh:195]: already exit in doc now.

File: examples/basic-tls/tidb-cluster.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (dragonly) [reply examples/basic-tls/tidb-cluster.yaml:57]: done

File: hack/gen-tls-certs.sh (hunk 4 above)
Comment (dragonly) [reply hack/gen-tls-certs.sh]: @csuzhangxc PTAL

File: examples/basic-tls/tidb-cluster.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply examples/basic-tls/tidb-cluster.yaml]: @csuzhangxc @dragonly Sorry that I missed this comment, if we set this, it means that users cannot connect to TiDB service with `mysql -h host -u user -p` command without the certs parameters, right? I think this is not necessary, I do not see any user configuring this till now.

**Comment by DanielZhangQD** [comment]
> > Do we need a command to delete all secrets after tested?
> 
> Thanks for reminding. I can add a `clean` subcommand in the script.

Is this done?

File: examples/basic-tls/tidb-cluster.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply examples/basic-tls/tidb-cluster.yaml]: > if we set this, it means that users cannot connect to TiDB service with mysql -h host -u user -p command without the certs parameters, right? 

YES. if no user need this, we may revert it back? @dragonly

**Review by csuzhangxc — APPROVED** [review]
LGTM execpt for https://github.com/pingcap/tidb-operator/pull/3594#discussion_r548775876

File: examples/basic-tls/tidb-cluster.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (dragonly) [reply examples/basic-tls/tidb-cluster.yaml]: reverted

> > > Do we need a command to delete all secrets after tested?
> > 
> > 
> > Thanks for reminding. I can add a `clean` subcommand in the script.
> 
> Is this done?

Yes, PTAL at https://github.com/pingcap/tidb-operator/pull/3594/files#diff-c5f3a75b252c9a0afe900c8ceb1af8915b20335daf8674d6baa1b7a91a7ba9efR198

**Comment by dragonly** [comment]
> > > Do we need a command to delete all secrets after tested?
> > 
> > 
> > Thanks for reminding. I can add a `clean` subcommand in the script.
> 
> Is this done?

Yes, PTAL at https://github.com/pingcap/tidb-operator/pull/3594/files#diff-c5f3a75b252c9a0afe900c8ceb1af8915b20335daf8674d6baa1b7a91a7ba9efR198

**Comment by dragonly** [comment]
> LGTM execpt for [#3594 (comment)](https://github.com/pingcap/tidb-operator/pull/3594#discussion_r548775876)

fixed

**Review by csuzhangxc — COMMENTED** [review]
LGTM

**Pull request merged**